source/credentials_provider_ecs.c

```diff
@@ -629,3 +628,93 @@ struct aws_credentials_provider *aws_credentials_provider_new_ecs(
+    struct aws_string *full_uri_str = NULL;
+    struct aws_uri full_uri;
+    AWS_ZERO_STRUCT(full_uri);
+    struct aws_credentials_provider *provider = NULL;
+
+    /* Read the auth token variables */
+    struct aws_string *ecs_env_token_file_path = NULL;
+    struct aws_string *ecs_env_token = NULL;
+    if (aws_get_environment_value(allocator, s_ecs_creds_env_token_file, &ecs_env_token_file_path) != AWS_OP_SUCCESS ||
+        aws_get_environment_value(allocator, s_ecs_creds_env_token, &ecs_env_token) != AWS_OP_SUCCESS) {
+        goto cleanup;
+    }
+    if (ecs_env_token_file_path != NULL) {
+        explicit_options.auth_token_file_path = aws_byte_cursor_from_string(ecs_env_token_file_path);
+    }
+    if (ecs_env_token != NULL) {
+        explicit_options.auth_token = aws_byte_cursor_from_string(ecs_env_token);
+    }
+
+    if (aws_get_environment_value(allocator, s_ecs_creds_env_relative_uri, &relative_uri_str) == AWS_OP_SUCCESS &&
+        relative_uri_str != NULL && relative_uri_str->len != 0) {
+
+        /* Using RELATIVE_URI */
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "ECS provider: using relative uri %s",
+            aws_string_c_str(relative_uri_str));
+
+        explicit_options.path_and_query = aws_byte_cursor_from_string(relative_uri_str);
+        explicit_options.host = aws_byte_cursor_from_string(s_ecs_host);
+        explicit_options.port = 80;
+
+        provider = aws_credentials_provider_new_ecs(allocator, &explicit_options);
+
+    } else if (
+        aws_get_environment_value(allocator, s_ecs_creds_env_full_uri, &full_uri_str) == AWS_OP_SUCCESS &&
+        full_uri_str != NULL && full_uri_str->len != 0) {
+
+        /* Using FULL_URI */
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "ECS provider: using full uri %s", aws_string_c_str(full_uri_str));
+        struct aws_byte_cursor full_uri_cursor = aws_byte_cursor_from_string(full_uri_str);
+        if (aws_uri_init_parse(&full_uri, allocator, &full_uri_cursor)) {
+            goto cleanup;
+        }
+
+        explicit_options.host = *aws_uri_host_name(&full_uri);
+        explicit_options.path_and_query = *aws_uri_path_and_query(&full_uri);
+        if (explicit_options.path_and_query.len == 0) {
+            explicit_options.path_and_query = aws_byte_cursor_from_c_str("/");
+        }
+
+        if (aws_byte_cursor_eq_c_str_ignore_case(aws_uri_scheme(&full_uri), "https")) {
+            explicit_options.tls_ctx = options->tls_ctx;
+        }
+
+        explicit_options.port = aws_uri_port(&full_uri);
+        if (explicit_options.port == 0) {
+            explicit_options.port = explicit_options.tls_ctx ? 443 : 80;
+        }
+
+        provider = aws_credentials_provider_new_ecs(allocator, &explicit_options);
+
+    } else {
+        /* Neither environment variable is set */
+        aws_raise_error(AWS_AUTH_CREDENTIALS_PROVIDER_INVALID_ENVIRONMENT);
```

Comment:
Current string for this error is [Valid credentials could not be sourced from process environment](https://github.com/awslabs/aws-c-auth/blob/277c8fbcb3f5341adb51000624b55ace301cfe82/source/auth.c#L42-L43)

Either make this string more generic, or add a more specific error-code

Reply:
trivial: Also, LOGF about this, at DEBUG or TRACE level. We don't need it polluting logs when it happens from default credentials provider, but if someone was calling it directly, it would be nice to see **something**

Reply:
I have made the AWS_AUTH_CREDENTIALS_PROVIDER_INVALID_ENVIRONMENT error more generic. I have already added a log in part 2, so not adding it here.